Add open3d::geometry::Image <-> sensor_msgs::msg::Image conversions

alexnavtt opened the pull request:
I'm currently working on a project using Open3D's TSDF integration pipeline and had a need to convert between ROS and Open3D image formats. I figured it might be worth submitting my conversion code here. This PR adds support for :
- `open3d::geometry::Image` -> `sensor_msgs::msg::Image`
- `open3d::camera::PinholeCameraIntrinsic` -> `sensor_msgs::msg::CameraInfo`
- `sensor_msgs::msg::Image` -> `open3d::geometry::Image`
- `sensor_msgs::msg::CameraInfo` -> `open3d::camera::PinholeCameraIntrinsic`

Additionally, since the data layout is the same in both formats, it provides move-only versions of 
- `open3d::geometry::Image` -> `sensor_msgs::msg::Image`
- `sensor_msgs::msg::Image` -> `open3d::geometry::Image`

I've run some qualitative tests using visualization in RViz and the Open3D visualizer to verify that images appear the same passing from ROS to Open3D and then back to ROS. I've performed similar tests for the CameraInfo functions. The formats I had available to me for testing were `rgb8` (RGB image), `16UC1` (depth image), and `mono8` (grayscale image)

alexnavtt committed:
Added conversion between open3d image and sensor_msgs image

alexnavtt committed:
Added camera info conversions and move conversions

alexnavtt committed:
Updated README for image conversions

SteveMacenski commented on open3d_conversions/include/open3d_conversions/open3d_conversions.hpp:
```diff
@@ -43,6 +45,38 @@ void open3dToRos(
+  const open3d::geometry::Image & o3d_img,
+  sensor_msgs::msg::Image & ros_img,
+  std::string encoding,
+  std::string frame_id = "open3d_image");
```
I'm not sure there's any rational default for this, I think it should be a required field (for all)

alexnavtt replied:
I agree. I had mainly been trying to match the functions that already existed. My thoughts are to remove that parameter entirely since a user would still have to assign the header timestamp field manually, so they might as well assign the frame_id manually as well.

SteveMacenski commented on open3d_conversions/src/open3d_conversions.cpp:
```diff
@@ -15,9 +15,45 @@
 #include "open3d_conversions/open3d_conversions.hpp"
 
+// This is the best we have prior to C++20 std::endian::native
+static bool isLittleEndian(){
```
A few tests for all these functions like in https://github.com/ros-perception/perception_open3d/blob/7cdffdfaa22f061b7652e4d2d6c34a39fdc9bea9/open3d_conversions/test/test_open3d_conversions.cpp are necessary to make sure there aren't any issues

SteveMacenski commented on open3d_conversions/src/open3d_conversions.cpp:
```diff
@@ -15,9 +15,45 @@
+  const int32_t value = 0x01;
+  const std::byte * least_significant_address = 
+    reinterpret_cast<const std::byte *>(&value);
+  return (*least_significant_address == std::byte{0x01});
```
Curious: where did you get this code from or did you write this? Just something I can glance over and assume is right if you grabbed it from somewhere else in ROS

alexnavtt replied:
I had found this code on StackOverflow as a non-UB version of checking endianess, but your comment here encouraged me to check how they do it in the cv_bridge package and apparently there's a `<rcpputils/endian.hpp>` header which behaves very similarly to C++ 20 std::endian, so I'll replace this function with that.

SteveMacenski reviewed (changes requested):
Make sure to lint the code with `ament_cpplint` and `ament_uncrustify`, I see a number of problems

alexnavtt committed:
Formatting changes after running ament_cpplint and ament_uncrustify

alexnavtt committed:
Switched to using rcpputils for checking endianness

alexnavtt committed:
Fixed spacing errors

alexnavtt committed:
Added endianness correction moving from ros to open3d image

alexnavtt committed:
Added test cases for image conversions

alexnavtt committed:
Removed frame_id inputs from all functions

alexnavtt committed:
Filled in copyright statement comment lines to fix failing copyright test

alexnavtt commented:
Alright, I've made all the changes including adding 8 new test cases, and I fixed one or two extra formatting issues that were causing `colcon test --ctest-args tests --packages-select open3d_conversions` to fail. I also realized that open3d requires its images to be in native endianness, so I added logic in the ros -> open3d conversion to flip the byte order if necessary.

SteveMacenski commented on open3d_conversions/include/open3d_conversions/open3d_conversions.hpp:
```diff
@@ -41,8 +43,37 @@ namespace open3d_conversions
  */
 void open3dToRos(
```
Ref in the comments that the user is expected to fill in the header afterwards with the frame/stamp

SteveMacenski reviewed (approved):
Otherwise generally looks good to me!

SteveMacenski commented:
@nkhedekar can you review?

alexnavtt committed:
Adjusted comments to account for the fact that frame_id is no longer used

alexnavtt commented:
Just checking back in. Maybe there's someone else who can review?

SteveMacenski commented:
@nkhedekar one more ping - can you review?

else @alexnavtt I can do a second look next week if you ping me back and we can take from there

nkhedekar commented:
Sorry its taken so long for a check. 
Thank you @alexnavtt for the contribution. Changes look good to me :+1:
